### PR TITLE
[MXNet-1340][Fit API]Update train stats

### PR DESCRIPTION
**Pull request by roywei**
## Description ##
In the previous [Fit-API design doc](https://cwiki.apache.org/confluence/display/MXNET/Gluon+Fit+API+-+Tech+Design#GluonFitAPI-TechDesign-Estimatorclass) , training statistics was stored in a dictionary,  some values are stored as a list such learning rates and training accuraccy over epochs. users has to understand the underlying data structure inorder to  access the statistics. 
This PR improves how event handlers access train stats, and reduced empty method call in event handlers to improve efficiency.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-1340](https://issues.apache.org/jira/browse/MXNET-1340) created
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/gluon/estimator/estimator.py [hunk 1]
```diff
@@ -236,32 +231,128 @@ def fit(self, train_data,
...
-            if self.stop_training:
+            if self.train_history.stop_training:
                 break
 
         # train end
         for handler in event_handlers:
             handler.train_end()
+
+
+class TrainHistory(object):
```
Comment (nswamy) [inline python/mxnet/gluon/estimator/estimator.py]: I think you should make this extensible without needing to derive this class for additional stats?

File: python/mxnet/gluon/trainer.py [hunk 2]
```diff
@@ -259,6 +259,13 @@ def learning_rate(self):
         else:
             return self._optimizer.learning_rate
 
+    @property
+    def optimizer(self):
+        if isinstance(self._optimizer, opt.Optimizer):
+            return self._optimizer
+        else:
+            raise UserWarning("Optimizer has not been initialized yet")
```
Comment (piyushghai) [inline python/mxnet/gluon/trainer.py]: What if the user sets a custom optimizer here?

File: python/mxnet/gluon/estimator/estimator.py [hunk 3]
```diff
@@ -125,16 +113,18 @@ def __init__(self, net,
...
+        # handle trainer
+        if not trainer:
             warnings.warn("No trainer specified, default SGD optimizer "
                           "with learning rate 0.001 is used.")
-            self.trainers = [gluon.Trainer(self.net.collect_params(),
-                                           'sgd', {'learning_rate': 0.001})]
+            self.trainer = gluon.Trainer(self.net.collect_params(),
+                                           'sgd', {'learning_rate': 0.001})
+        elif not isinstance(trainer, gluon.Trainer):
+            raise ValueError("Trainer must be a Gluon Trainer instance, refer to gluon.trainer")
```
Comment (piyushghai) [inline python/mxnet/gluon/estimator/estimator.py]: Since you say refer to gluon.trainer, you could probably also add a url to gluon.trainer docs : http://mxnet.incubator.apache.org/versions/master/api/python/gluon/gluon.html#trainer

File: python/mxnet/gluon/estimator/estimator.py [hunk 4]
```diff
@@ -236,32 +231,128 @@ def fit(self, train_data,
...
-                    self.train_stats['batch_' + loss_metric.name] = loss_metric.get()[1]
+                    self.train_history.set_batch_stats(*loss_metric.get())
 
-                try:
-                    self.train_stats['step'] = "{}/{}".format(batch_size * (i + 1), len(train_data._dataset))
-                except AttributeError:
-                    self.train_stats['step'] = i
+                self.train_history.batch_idx = i
+                # record trained samples v.s. total samples if using Gluon DataLoader
+                if isinstance(train_data, gluon.data.DataLoader):
```
Comment (piyushghai) [inline python/mxnet/gluon/estimator/estimator.py]: You might need to rebase this line with the fit-api branch again. 
https://github.com/apache/incubator-mxnet/blob/fit-api/python/mxnet/gluon/estimator/estimator.py#L245

**Comment by pinaraws** [comment]
@mxnet-label-bot add[pr-awaiting-review, Gluon]

File: python/mxnet/gluon/trainer.py (hunk 2 above)
Comment (roywei) [reply python/mxnet/gluon/trainer.py]: a custom optimizer should still inherit the base Optimizer class. Gluon trainer does the check: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/trainer.py#L123

File: python/mxnet/gluon/estimator/estimator.py [hunk 5]
```diff
@@ -231,37 +221,45 @@ def fit(self, train_data,
...
+            event_handlers.append(LoggingHandler())
 
-        # training begin
+        train_begin, epoch_begin, batch_begin, \
+        batch_end, epoch_end, train_end = self._categorize_handlers(event_handlers)
+
+        # passing estimator to event handlers so they can access estimator information
+        # when a event is triggered
         for handler in event_handlers:
+            handler.estimator = self
```
Comment (roywei) [inline python/mxnet/gluon/estimator/estimator.py]: @nswamy This will avoid to ask user passing estimator during event handler construction, reference: https://github.com/apache/incubator-mxnet/pull/14462#discussion_r267125641

File: python/mxnet/gluon/estimator/estimator.py [hunk 6]
```diff
@@ -39,27 +39,26 @@ class Estimator(object):
...
     def __init__(self, net,
-                 loss=None,
+                 loss,
                  metrics=None,
                  initializer=None,
-                 trainers=None,
+                 trainer=None,
                  context=None):
 
         self.net = net
```
Comment (nswamy) [inline python/mxnet/gluon/estimator/estimator.py]: do you want to set self._estimator = None?

File: python/mxnet/gluon/estimator/estimator.py (hunk 5 above)
Comment (nswamy) [reply python/mxnet/gluon/estimator/estimator.py]: I am wondering how the user of handler will know that an estimator will be initialized here? Also can you have a setter and getter for the estimator in Handler and not call handler.setEstimator(e) if handler.getEstimator() is not None.

Comment (roywei) [reply python/mxnet/gluon/estimator/estimator.py]: when user call est.fit(xxx, event_handlers=XXX), this will already associate the event handlers with an estimator instance. I m just helping the user to pass this estimator so they don't need to do so during event handler construction. 
The getter and setter are already implemented through the property interface. `handler.estimator=self` is actually the setter method of property estimator.

File: python/mxnet/gluon/estimator/estimator.py (hunk 6 above)
Comment (roywei) [reply python/mxnet/gluon/estimator/estimator.py]: This is the estimator class, only event handlers should have self._estimator?

File: python/mxnet/gluon/estimator/estimator.py [hunk 7]
```diff
@@ -298,42 +299,65 @@ def fit(self, train_data,
...
 
             if self.stop_training:
                 break
 
         # train end
-        for handler in event_handlers:
+        for handler in train_end:
             handler.train_end()
+
+    def categorize_handlers(self, event_handlers):
```
[the remark this answers is not in the document]
Comment (roywei) [reply python/mxnet/gluon/estimator/estimator.py]: updated

File: python/mxnet/gluon/estimator/estimator.py [hunk 8]
```diff
@@ -298,42 +299,65 @@ def fit(self, train_data,
...
+        implemented will be called at train begin
+        """
+
+        train_begin = []
+        epoch_begin = []
+        batch_begin = []
+        batch_end = []
+        epoch_end = []
+        train_end = []
+        base_handler = EventHandler()
```
[the remark this answers is not in the document]
Comment (roywei) [reply python/mxnet/gluon/estimator/estimator.py]: removed

**Comment by roywei** [comment]
@nswamy I have addressed the comments, could you take another look? thanks!

**Comment by piyushghai** [comment]
@roywei Can you look at the CI failures ?

**Comment by roywei** [comment]
@piyushghai it's all due to R package failure, i think it's ok. We will rebase before merging to master, and hopefully the R test will pass.
